### DISCUSSION TITLE
Multiple CIDRs per block and API

### DISCUSSION THREAD

**Discussion by whizzer929** [Azure/ipam#68] [General]
Hello,

We're evaluating your IPAM and so far it's working well.  However, we have a use case where we have multiple CIDR subnets that would make up a block.  I.E.  space = dev, block = private, CIDRs = 10.40.0.0/18 AND 10.40.64.0/19.  Our only work around is too create multiple blocks with index number.  private_1, private_2.  However, this now dictates the use of multiple API URLs and choosing which one to use is problematic.

How can we have multiple associated blocks and a single API URL?

Thanks.

**Comment by DCMattyG** [comment]
Hi @whizzer929, appreciate you reaching out!

The important thing to keep in mind is that a Block represents a CIDR space, and any vNET's (subnets) that are a part of that block must fit within the Block's CIDR.

In the above example, I believe you were stating that you have two different CIDR's applied to a vNET (10.40.0.0/18 and 10.40.64.0/19), correct? What was the Block CIDR you assigned (if I may ask)?

P.S. If I misunderstood anything, please keep me honest. Thanks!

**Reply by whizzer929** [reply]
Hi DCMattyG,

Yes, in our case we have multiple separate, non-overlapping CIDRs blocks that would belong to a specific classification that we define.  I.E.  (Dev/Private), (Dev/Public), (NonProd/Private), etc.  I.E. 10.40.0.0/18 and 10.40.64.0/19 would both be classified as (Dev/Private).

Our initial plan was to assign environments (dev, nonprod, prod) to Spaces, and private, public to blocks.  So the Dev space, would have a public and private block.  But since we have more than one CIDRs block classified as Dev/Private, we would have to create multiple blocks to encompass all the IP space.  I.E. Dev/Private_1, Dev/Private_2.  The problem comes with the fact the block name makes up the URL to the API, specifically the reservations API.

So our use case is, just give me the next reservation of size /xx from Dev/Private.  I do not want to have to be concerned with calling one of the two APIs URLs; https://ipmadev.azurewebsites.net/api/spaces/dev/blocks/private_1/reservations or https://ipmadev.azurewebsites.net/api/spaces/dev/blocks/private_2/reservations.  I would want to have one URL to call.

Hope this explains it enough.  Thanks for you prompt attention.

**Reply by DCMattyG** [reply]
Got it, I'm on the same page!

Would it make sense to assign a larger CIDR to the Block, something like 10.40.0.0/17 perhaps, or you need to have exactly the above space, and nothing more in your Block definition? (Just asking questions for additional clarity)

**Reply by whizzer929** [reply]
Unfortunately our IP space isn't carved out in nice clean boundaries.  In this case, 10.40.96.0/19 belongs to another another class.  I'm not sure there is an answer for this use case, maybe we can flag this as a feature request?  I.E. Add multiple subnets to a block.

**Reply by DCMattyG** [reply]
Funny story @whizzer929, I'm actually on the phone with @amitjattan right now talking about this exact issue for you. He's trying to pull you into the meeting haha!

**Reply by whizzer929** [reply]
Yeah, we work together.  lol.  I never imagined you'd be speaking with him.

**Reply by DCMattyG** [reply]
After discussing this @amitjattan, it seems like the only path forward would be a feature request to have multiple, non-summarize-able CIDR's as part of a Block. I'm still not quite sure what that will look like in my head hahaha, but the only other option I can think of is perhaps some kind of "Batch" API where you can provide multiple Blocks and get back the first new CIDR of the target/requested size. 

I have a feeling the latter would be far easier to implement as it wouldn't require core changes in the design, but I'm always open to your thoughts on if this would completely meet your needs if that's the path we go down.

Looking forward to your feedback!

**Reply by whizzer929** [reply]
"some kind of "Batch" API where you can provide multiple Blocks and get back the first new CIDR of the target/requested size."  That should work.

**Reply by DCMattyG** [reply]
I'm on it!

**Reply by DCMattyG** [reply]
Good morning @whizzer929, I have the draft of this feature built and added to the IPAMDEV containers. Please give it a try and let me know if you have any feedback/improvements.

*Assuming* you are using the standard Full deployment, you can switch to the DEV containers by updating the Deployment Settings in your App Service as follows:

![image](https://user-images.githubusercontent.com/12955732/213220881-181f56e1-fc9b-4838-86c9-e1b94ac0c832.png)

Once you have that saved & restarted, you'll find the new API in the Swagger Docs:

![image](https://user-images.githubusercontent.com/12955732/213221061-e384b315-6fd0-4f9f-ac67-23d8a097619e.png)

This API will be called at the *Space* level, and is essentially the same as the reservations API at the *Block* level, except this takes a list (array) of Blocks from within the Space as a parameter, and will evaluate them ***in the order they are provided***.

**EXAMPLE BODY**

```
{
    "blocks": ['TestBlock01', 'TestBlock02'],
    "size": 24
}
```

Hope you enjoy, please let me know how you like it!

**Comment by whizzer929** [comment]
Hi DCMattyG, we just got the new code and will start testing.  So far, it looks good.  Thanks for the update!  Your responsiveness is much appreciated.
